Change de-identification method for Sensitive ID export fields

nospame opened the pull request:
## Product Description
This will result in a one-time shift of values for de-identified "Sensitive ID" fields in exports.

## Technical Summary
[SAAS-16366](https://dimagi.atlassian.net/browse/SAAS-16366)
The main concept here is that, instead of using a consistent hashing algorithm for DeID'd fields, we'll create and store a UUID that can be looked up for any given domain, which will be used to return a hash of the input value and the UUID. This way, the DeID'd output will still stay consistent over time but cannot be determined by simply re-running the hashing process on a value. This is represented by `DeIdHash`.

In order to add a model in the `export` app that could be used in the same place as other transform functions, I had to tweak how we access those functions in `export/const.py`. Previously these functions were all imported directly, but (perhaps predictably), adding a reference to a model from a django app within that app's own constants, would have resulted in circular imports. This is dealt with by adding `get_transform_function` and `get_deid_transform_function` to `export/utils.py`.


## Safety Assurance

### Safety story
This is a change that interacts with core export functionality, so there is some risk here. I've tested locally to see that marked fields are DeID'd correctly and that there is no other change to given exports. There is decent automated test coverage of the `export` app as a whole and would expect that to catch any major functional issues.

### Automated test coverage
Added a small test suite for behavior of the `get_deid` method of `DeIdHash`. The `DEID_ID_TRANSFORM` that now points to the new method is implicitly tested in a few places throughout `hqcase/tests`.

### QA Plan
Technically this is not a huge change, but given the risk in changes to exports, I think it'll be worthwhile to have this get a pass by QA.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations
The migration would need to be reverted first.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-16366]: https://dimagi.atlassian.net/browse/SAAS-16366?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Get export transform functions with importlib via util

nospame committed:
Add DeIdMapping

nospame committed:
Point get_deid_transform_function to DeIdMapping

nospame committed:
Add TestDeIdMapping

nospame committed:
Return a string from get_deid

Ensures it can be transformed to xml

nospame committed:
Add DeIdMapping to domain dump and deletion models

nospame commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+
+class DeIdMapping(models.Model):
+    domain = models.TextField(max_length=255)
+    hashed_value = models.TextField(max_length=32)
+    deid = models.UUIDField(default=uuid.uuid4)
```
I had trouble coming up with good names for these fields, with fairly little context on export code in general. Very open to alternatives here. Maybe something closer to `hashed_item` and `deid_value`?

nospame commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+    @staticmethod
+    @quickcache(['value'], timeout=90 * ONE_DAY)
+    def _hash_value(value):
+        if value is None:
+            # None is a de-identifiable value but needs a string to encode for lookup
+            value = ''
```
I don't think there's a meaningful different between an empty string and `None` in data this would be de-identifying.

Previously you would get the same output for the both `None` and the string `'None'` and this seems better to me, but if there's any reason to retain the previous behavior I can just call `str(value)` instead.

nospame committed:
Always get domain from doc or case for DeIdMapping

nospame commented:
@coderabbitai review

millerdev commented on corehq/apps/export/utils.py:
```diff
@@ -57,10 +57,21 @@ def get_default_export_settings_if_available(domain):
+
+
+def get_transform_function(func_name):
+    module = importlib.import_module('corehq.apps.export.transforms')
```
Why is `importlib` necessary here? Does a normal import not work?
```suggestion
    import corehq.apps.export.transforms as module
```

millerdev commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+        return cls._get_deid(value, domain)
+
+    @classmethod
+    @quickcache(['value', 'domain'], timeout=90 * ONE_DAY)
```
90 days is a long time. If the code behind this cache changes, it will not take effect until months later (for cached values). Can you explain more about why such a long cache timeout was chosen?

nospame replied:
I hadn't considered code changes effectively being delayed by cache, which is a good point, but the length of time is because these values are essentially expected to stay the same forever once stored. The reason for caching here generally is so we're not making excessive DB calls for every repeated "sensitive ID" value in an export.

Maybe there's a shorter timeout here that would be more convenient for consideration of working with this code in the future, but still mitigate any performance impact on exports. Maybe 1 month? 1 week?

millerdev replied:
I think even one day or even one hour would be fine if the intent is to reduce database load.

millerdev commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+        return deid_mapping.deid
+
+    @staticmethod
+    @quickcache(['value'], timeout=90 * ONE_DAY)
```
Why does the result of this function need to be cached?

My thoughts: The cache adds a network hit (to redis), but the function itself does not hit the network, so the cache would most likely make the function slower. Also, MD5 hashing should be very fast unless `value` is huge, but even in that case saving a huge value to the cache could be problematic.

nospame replied:
I cached it here without thinking about the additional network hit, just that the return value will stay the same for a given input so we "might as well" cache it. I assume you're probably correct that simply hashing the value and returning it would be more performant than caching it here, so I'll make that change.

millerdev commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+from corehq.util.quickcache import quickcache
+
+
+class DeIdMapping(models.Model):
```
Do I have it right that the purpose of this model is to map values to UUIDs so that a given value always has the same UUID?

Could the model be eliminated by using the hash of the value + domain, possibly with a stronger hash function (SHA-256, for example)? That would make the `get_deid` function look something like:

```py
from haslib import sha256

def get_deid(value, doc):
    domain = doc["domain"] if isinstance(doc, dict) else doc.domain
    data = f"{value} {domain}".encode("utf-8")
    return sha256(data).hexdigest()[:32]
```

We could also use something like a secret salt (possibly stored in settings.py, or even one per domain) to make it harder to re-identify the values, although that may be overkill.

millerdev replied:
Yet another idea is to use the document ID as hash salt.

nospame replied:
> Do I have it right that the purpose of this model is to map values to UUIDs so that a given value always has the same UUID?

That's essentially right, so that in exported data for a given project space, the same value is always de-identified to the same output.

> Could the model be eliminated by using the hash of the value + domain, possibly with a stronger hash function (SHA-256, for example)? 

SHA-256 will always return the same value given the same inputs, right? The other point in looking at a stored random value over a function is that it is not reproducible -- that is, someone could not (theoretically) run our code on a set of raw values, record the de-identified outputs, and compare that with the values shown in a de-identified export. That potential concern is the primary reason for moving away from the existing de-identification method.

Would a secret salt deal with that concern? Do you know how something like that could be implemented? We wouldn't want to use the document ID, because the output should be consistent across the same value occurring in any export on the project space, regardless of what document it originally came from.

Given the reasons above, it honestly just seemed simpler to use a db model here, but I'm certainly not a cryptography expert.

nospame replied:
> even one per domain

Thinking through it, could this potentially be a model storing a UUID as a salt for a given domain, rather than storing the de-identified values themselves? Thinking that would look more like this:

```python
class DeIdMapping(models.Model):
    domain = models.TextField(max_length=255)
    salt = models.UUIDField(default=uuid.uuid4)

    @classmethod
    def get_deid(value, doc):
        domain = doc["domain"] if isinstance(doc, dict) else doc.domain
        mapping, __ = cls.objects.get_or_create(domain=domain)
        data = f"{value} {mapping.salt}".encode("utf-8")
        return sha256(data).hexdigest()[:32]
```

Would think a bit on naming as well, not sure if `DeIdMapping` still fits exactly, in that case.

millerdev replied:
Yes, that's what I had in mind with the "one per domain" idea.

nospame replied:
Left it at md5 because we're not concerned with _security_ here, just general data privacy, and I believe it's a good bit faster than sha256, which is probably relevant for those hundred-thousand-row exports. https://github.com/dimagi/commcare-hq/pull/35772/commits/2abec349ec981a164354da718094970cae4fb57b

millerdev commented on corehq/apps/export/migrations/0014_deidmapping.py:
```diff
@@ -0,0 +1,39 @@
+            fields=[
+                (
+                    "id",
+                    models.AutoField(
```
How many of these do we expect to have? If the number grows very large, especially if there is a lot of churn, would it be worth using `BigAutoField`?

nospame replied:
I don't have a sense of how widely the de-identified data feature is used, but it seems plausible we could run up against the limit using if a regular `AutoField`. If we consider exports from our larger projects that might have several hundred thousand unique cases, if each of those has a few different de-identified fields, and a few different projects of that scale use this feature, it would add up pretty quickly, so `BigAutoField` is likely worthwhile.

millerdev replied:
I feel like this suggestion of using `BigAutoField` may have pointed in a direction I'm not really in favor of. Thinking about this more, I'm strongly in favor of using a different hashing technique that doesn't require a new database model, such as using a unique salt per domain. Yet another idea is to use the document ID as hash salt.

millerdev commented on corehq/apps/export/utils.py:
```diff
@@ -57,10 +60,25 @@ def get_default_export_settings_if_available(domain):
+
+def get_deid_transform_function(func_name):
+    if func_name == DEID_TRANSFORM_FUNCTIONS[DEID_DATE_TRANSFORM]:
+        module = importlib.import_module('couchexport.deid')
```
```suggestion
        import couchexport.deid as module
```

millerdev commented on corehq/apps/export/utils.py:
```diff
@@ -73,5 +76,9 @@ def get_transform_function(func_name):
+        return getattr(module, func_name)
+    elif func_name == DEID_TRANSFORM_FUNCTIONS[DEID_ID_TRANSFORM]:
+        from corehq.apps.export.models import DeIdMapping
+        return DeIdMapping.get_deid
```
This function will return `None` when neither of the if-conditions are met. Is that intentional?

nospame replied:
`deid_transform` that ultimately gets passed in here is set from the list of `DEID_TRANSFORM_FUNCTIONS` as choices:
https://github.com/dimagi/commcare-hq/blob/1a32a6bc9481f31cb8212f7e207b9d33eae1db23/corehq/apps/export/models/new.py#L278

If what gets passed in here is not in that list of choices, something has gone wrong, but I wouldn't expect that to be possible. The caller would then error on trying to call `None` as a function. Is that a behavior worth writing a test for?

nospame committed:
Don't use importlib for get_transform functions

nospame committed:
Use BigAutoField

gherceg commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -0,0 +1,42 @@
+class DeIdMapping(models.Model):
+    domain = models.TextField(max_length=255)
+    hashed_value = models.TextField(max_length=32)
+    deid = models.UUIDField(default=uuid.uuid4)
```
If we do stick with the db model, could we just make the primary key a uuid and use that instead of the deid field? (Code might not be 100% accurate)
`id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)`

millerdev replied:
+1

gherceg commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -18,11 +18,11 @@ class Meta:
+        if isinstance(doc, dict):
             domain = doc['domain']
-
+        else:
+            domain = doc.domain
```
Probably good to have a test cover this code as well.

nospame replied:
https://github.com/dimagi/commcare-hq/pull/35772/commits/918c24a426bf239b8cdc5a58d1ef900a2a5e47f5

gherceg reviewed (commented):
Pair reviewed with @jingcheng16. Curious to hear the response to Daniel's feedback about using a sha256 hash function instead of a db model.

nospame committed:
DeIdMapping -> DeIdHash

Main changes:
 - store 1 row per domain, rather than each domain/value combo
 - hash it live using the stored UUID as a salt

nospame committed:
Update migrations.lock

nospame committed:
Add test_uses_domain_from_dict_or_object

nospame commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -1,42 +1,24 @@
+    @quickcache(['domain'], timeout=24 * 60 * 60)
+    def _get_salt(cls, domain):
+        instance, __ = cls.objects.get_or_create(domain=domain)
+        return instance.id
```
Good call. It does work as is, but better to be explicit about wanting a string here.

nospame committed:
Use id.hex to return a string instead of UUID object

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

nospame committed:
Use original output format for de-identified value

nospame commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -14,8 +15,7 @@ class DeIdHash(models.Model):
         salt = cls._get_salt(domain)
-        data = f'{value}{salt}'.encode('utf-8')
-        return md5(data).hexdigest()
+        return DeidGenerator(value, salt).random_hash()
```
Got some feedback on the ticket that we should retain the original 10-character output format for these values. Just using the previous generator (with the addition of the new unique salt) felt like the nicest way to do that without reinventing the wheel.

millerdev replied:
Would `md5(data).hexdigest()[:10]` also work? From my perspective it's far simpler.

nospame replied:
Not quite, since it's only hex (0-9, A-F) versus base 36 (0-9, A-Z). It is a subset of the expected format and would technically _work_ but I believe the possibility for hash collision would be much higher. I looked into approaches to get a hex or bytes hash into a base 36, and everything I found seemed like it would be essentially reinventing most of what `DeidGenerator` is already doing.

millerdev replied:
Does a hash collision matter here? I think it may not.

For small numbers of hashes (n<1000), the probability is very low. For large numbers (approaching 1 million, for example), a collision does get significantly more likely (50%).

millerdev replied:
I think occasional collisions probably do not matter in the case of DeID. Practically it means that two pieces of sensitive information that previously did not match could end up with the same value. But the object of DeID is to make sensitive information impossible to guess, and I think collisions actually help in that regard.

Another property of this DeID implementation is that common values (names like 'John' or an empty field) will always have the same hash for a given domain, and contextual information may make those values easier to guess. Having the possibility of collisions dilutes that ever so slightly.

nospame replied:
Couple of things here:
- got input from @kcowger that we do want to do our best to guarantee uniqueness of de-identified values even in very large exports.
- there is likely a bug with certain form exports in this code that I need to address, where either the `doc` passed in is missing or invalid for whatever reason and there's no `domain` to get from it. Based on current prioritization putting this on my back-burner, that'll likely get done mid-April.

millerdev reviewed (approved):
I wasn't able to quickly make sense of what `DeidGenerator` does, don't consider this approval as an endorsement of that class.

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/deid-export-method

nospame committed:
Simplify use of DEID constants

nospame committed:
Pass in domain to DeIdHash directly where possible

nospame committed:
Handle bug with bad data from deleted mobile repeat groups

nospame committed:
Test DeId output format

nospame commented on corehq/apps/export/models/deid_export.py:
```diff
@@ -12,8 +12,10 @@ class DeIdHash(models.Model):
-    def get_deid(cls, value, doc):
-        domain = doc['domain'] if isinstance(doc, dict) else doc.domain
+    def get_deid(cls, value, doc, domain=None):
+        domain = (domain if domain is not None
```
In digging into behavior with de-identifying data in repeat groups on form exports, I realized that the `doc` passed in for these fields is just a dict representation of the repeat itself, not an entire form or case as is typical, so they would have errored on trying to access `doc['domain']`. I'm passing in domain directly here now in most situations, except where it's used by the `CaseCopier`, because that should always have a `CommCareCase` to pass in.

That probably means I can remove the following line, but I sort of lean toward keeping in in to catch edge cases, because there's definitely funky data in some places in HQ.
```
else doc['domain'] if isinstance(doc, dict)
```
